app/Insightly/Listeners/SyncContact.php

```diff
@@ -0,0 +1,127 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Insightly\Listeners;
+
+use App\Domain\Contacts\Contact;
+use App\Domain\Contacts\Events\ContactCreated;
+use App\Domain\Contacts\Events\ContactUpdated;
+use App\Domain\Contacts\Repositories\ContactRepository;
+use App\Insightly\Exceptions\ContactCannotBeUnlinked;
+use App\Insightly\InsightlyClient;
+use App\Insightly\InsightlyMapping;
+use App\Insightly\Repositories\InsightlyMappingRepository;
+use App\Insightly\Resources\ResourceType;
+use App\Insightly\SyncIsAllowed;
+use Illuminate\Bus\Queueable;
+use Illuminate\Contracts\Queue\ShouldQueue;
+use Illuminate\Support\Arr;
+use Psr\Log\LoggerInterface;
+
+final class SyncContact implements ShouldQueue
+{
+    use Queueable;
+
+    public function __construct(
+        private readonly InsightlyClient $insightlyClient,
+        private readonly ContactRepository $contactRepository,
+        private readonly InsightlyMappingRepository $insightlyMappingRepository,
+        private readonly LoggerInterface $logger,
+    ) {
+    }
+
+    public function handleContactCreated(ContactCreated $contactCreated): void
+    {
+        $contact = $this->contactRepository->getById($contactCreated->id);
+        if (!SyncIsAllowed::forContact($contact)) {
+            return;
+        }
+
+        $this->storeAndLinkContactAtInsightly($contact);
+
+        $this->logger->info(
+            'Contact created',
+            [
+                'domain' => 'insightly',
+                'contact_id' => $contactCreated->id->toString(),
+            ]
+        );
+    }
+
+    private function storeAndLinkContactAtInsightly(Contact $contact): void
+    {
+        $contactIds = $this->insightlyClient->contacts()->findIdsByEmail($contact->email);
+
+        if (empty($contactIds)) {
+            $contactInsightlyId = $this->insightlyClient->contacts()->create($contact);
+        } else {
+            $contactIds = array_values(Arr::sort($contactIds));
+            $contactInsightlyId = $contactIds[0];
+
+            $this->insightlyClient->contacts()->update($contact, $contactInsightlyId);
+        }
+
+        $this->insightlyMappingRepository->save(new InsightlyMapping(
+            $contact->id,
+            $contactInsightlyId,
+            ResourceType::Contact
+        ));
+
+        $integrationMapping = $this->insightlyMappingRepository->getByIdAndType($contact->integrationId, ResourceType::Opportunity);
```

Comment:
With a recent change to the mapping table, I think we also need to make sure to update the project mapping if it exists.

Do we also need to make sure the mapping exists in the database?

Reply:
I don't really get why the opportunity mapping should be updated. The contact is linked or unlinked from an opportunity, but the opportunity itself doesn't change.

Reply:
What I meant was that an integration can now have an opportunity and a project inside Insightly. So now the contact is only linked to the opportunity but it should also be linked to the project.

So that is also why it is important to validate that the mapping exists, an integration could have an opportunity and/or project mapping. (When doing the migration of the old Project Aanvraag data it will even only have a project mapping)